File: core/src/main/java/com/terralite/core/registry/SimpleFrozenRegistry.java
package com.terralite.core.registry;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class SimpleFrozenRegistry<T> implements FrozenRegistry<T> {
    private final RegistryKey<T> key;
    private final Map<ResourceId, T> entries;

    public SimpleFrozenRegistry(RegistryKey<T> key, Map<ResourceId, T> entries) {
        this.key = Objects.requireNonNull(key, "key");
        this.entries = Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(entries, "entries")));
    }

    @Override
    public RegistryKey<T> key() {
        return key;
    }

    @Override
    public Optional<T> get(ResourceId id) {
        return Optional.ofNullable(entries.get(id));
    }

    @Override
    public T require(ResourceId id) {
        T value = entries.get(id);
        if (value == null) {
            throw new IllegalArgumentException("Missing registry entry: " + id + " in " + key.id());
        }
        return value;
    }

    @Override
    public boolean contains(ResourceId id) {
        return entries.containsKey(id);
    }

    @Override
    public Collection<ResourceId> ids() {
        return entries.keySet();
    }

    @Override
    public Collection<T> values() {
        return entries.values();
    }

    @Override
    public int size() {
        return entries.size();
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/SimpleFrozenRegistry.kt
package com.terralite.core.registry

/** Immutable snapshot created by [SimpleMutableRegistry.freeze]. */
class SimpleFrozenRegistry<T>(
    private val registryKey: RegistryKey<T>,
    entries: Map<ResourceId, T>
) : FrozenRegistry<T> {

    private val entries: Map<ResourceId, T> = LinkedHashMap(entries)

    override fun key(): RegistryKey<T> = registryKey

    override fun get(id: ResourceId): T? = entries[id]

    override fun require(id: ResourceId): T =
        entries[id] ?: throw IllegalArgumentException("Missing registry entry: $id in ${registryKey.id}")

    override fun contains(id: ResourceId): Boolean = entries.containsKey(id)

    override fun ids(): Collection<ResourceId> = entries.keys

    override fun values(): Collection<T> = entries.values

    override fun size(): Int = entries.size
}
